mirrord/protocol/src/codec.rs

```diff
@@ -142,77 +143,37 @@ pub enum DaemonMessage {
     SwitchProtocolVersionResponse(#[bincode(with_serde)] semver::Version),
 }
 
-pub struct ClientCodec {
+pub struct ProtocolCodec<I, O> {
     config: bincode::config::Configuration,
+    /// Phantom just to associate the message types with the struct.
+    /// Implementing a trait to associate the types is not good enough, because we can't blanket
+    /// implement the foreign `Encoder`/`Decoder` traits, so we have to implement them for a
+    /// struct.
+    _phantom_incoming_message: PhantomData<I>,
+    _phantom_outgoing_message: PhantomData<O>,
 }
 
-impl ClientCodec {
-    pub fn new() -> Self {
-        ClientCodec {
-            config: bincode::config::standard(),
-        }
-    }
-}
-
-impl Default for ClientCodec {
-    fn default() -> Self {
-        ClientCodec::new()
-    }
-}
-
-impl Decoder for ClientCodec {
-    type Item = DaemonMessage;
-    type Error = io::Error;
-
-    fn decode(&mut self, src: &mut BytesMut) -> io::Result<Option<Self::Item>> {
-        match bincode::decode_from_slice(&src[..], self.config) {
-            Ok((message, read)) => {
-                src.advance(read);
-                Ok(Some(message))
-            }
-            Err(DecodeError::UnexpectedEnd { .. }) => Ok(None),
-            Err(err) => Err(io::Error::new(io::ErrorKind::Other, err.to_string())),
-        }
-    }
-}
-
-impl Encoder<ClientMessage> for ClientCodec {
-    type Error = io::Error;
+pub type ClientCodec = ProtocolCodec<DaemonMessage, ClientMessage>;
```

Comment:
```suggestion
/// Type alias used to deal with the messages we receive from the agent. You'll see this being consumed by our hooks.
pub type ClientCodec = ProtocolCodec<DaemonMessage, ClientMessage>;
```

Reply:
Or something along those lines.

Reply:
I don't think that comment is correct. How about: 
```suggestion
// Codec to be used by the client side to receive `DaemonMessage`s from the agent and send `ClientMessage`s to the agent.
pub type ClientCodec = ProtocolCodec<DaemonMessage, ClientMessage>;
```